src/Kiota.Builder/Writers/TypeScript/CodeFunctionWriter.cs

```diff
@@ -95,36 +205,111 @@ private void WriteDiscriminatorFunction(CodeFunction codeElement, LanguageWriter
+        writer.WriteLine($"return {deserializationFunction.ToFirstCharacterLowerCase()}{parseNodeParameterForPrimitiveValues};");
+    }
 
-        return conventions.GetTypeString(new CodeType { TypeDefinition = parent }, codeElement, false);
+    private static bool ShouldWriteDiscriminatorInformation(CodeFunction codeElement, CodeComposedTypeBase? composedType)
+    {
+        return codeElement.OriginalMethodParentClass.DiscriminatorInformation.ShouldWriteDiscriminatorForInheritedType || composedType is CodeUnionType;
+    }
+
+    private void WriteDiscriminatorInformation(CodeFunction codeElement, CodeParameter parseNodeParameter, LanguageWriter writer)
+    {
+        writer.WriteLines($"const mappingValueNode = {parseNodeParameter.Name.ToFirstCharacterLowerCase()}.getChildNode(\"{codeElement.OriginalMethodParentClass.DiscriminatorInformation.DiscriminatorPropertyName}\");",
+                            "if (mappingValueNode) {");
+        writer.IncreaseIndent();
+        writer.WriteLines("const mappingValue = mappingValueNode.getStringValue();",
+                        "if (mappingValue) {");
+        writer.IncreaseIndent();
+
+        writer.StartBlock("switch (mappingValue) {");
+        foreach (var mappedType in codeElement.OriginalMethodParentClass.DiscriminatorInformation.DiscriminatorMappings)
+        {
+            writer.StartBlock($"case \"{mappedType.Key}\":");
+            writer.WriteLine($"return {GetFunctionName(codeElement, mappedType.Value.Name.ToFirstCharacterUpperCase(), CodeMethodKind.Deserializer)};");
+            writer.DecreaseIndent();
+        }
+        writer.CloseBlock();
+        writer.CloseBlock();
+        writer.CloseBlock();
+    }
+
+    private string GetParseNodeParameterForPrimitiveValues(CodeFunction codeElement, CodeParameter? parseNodeParameter)
+    {
+        if (GetOriginalComposedType(codeElement.OriginalLocalMethod.ReturnType) is CodeComposedTypeBase composedType && IsComposedOfPrimitives(composedType) && parseNodeParameter is not null)
+        {
+            return $"({parseNodeParameter.Name.ToFirstCharacterLowerCase()})";
+        }
+        return string.Empty;
+    }
+
+    private string GetFunctionName(CodeElement codeElement, string returnType, CodeMethodKind kind)
+    {
+        var functionName = GetFunctionName(returnType, kind);
+        var parentNamespace = codeElement.GetImmediateParentOfType<CodeNamespace>();
+        var codeFunction = FindCodeFunctionInParentNamespaces(functionName, parentNamespace);
+        return conventions.GetTypeString(new CodeType { TypeDefinition = codeFunction }, codeElement, false);
+    }
+
+    private CodeFunction? FindCodeFunctionInParentNamespaces(string functionName, CodeNamespace? parentNamespace)
+    {
+        CodeFunction? codeFunction;
+        do
+        {
+            codeFunction = parentNamespace?.FindChildByName<CodeFunction>(functionName);
```

Comment:
why do you need to iterate and check the name here since FindChildByName already does the name match for you?

Reply:
It fails Integration tests because it does case insensitive comparison: https://github.com/microsoft/kiota/actions/runs/9595099902/job/26459200361?pr=4602

Reply:
I'm really curious to why it'd find an equivalent with a different casing. To me it sounds like we'd done something twice for whatever reason, and forgot to remove the older element. Is the other function actually used anywhere? Please investigate.

Reply:
Okay i'll check each step to confirm